### PR TITLE
P2Plus: SourceAcademy Mission Creation Interface

### PR DESCRIPTION
**Pull request by ChengGeng97**
### Description
Creates a way for GitHub-hosted missions to be created through the Source Academy website.

Subfeatures:
- Added MCQ support for GitHub-hosted missions
- Ability to create and delete new tasks for a mission
- Ability to create, edit and delete test cases for a mission
- Ability to load test cases
- Control of admin features so that they are only accessible to teachers

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Code quality improvements

### How to test
To test the autotester and MCQ functionality (as a student):
1. Accept the assignment at this link: https://classroom.github.com/a/D16hWvjA
2. Open the mission in the SA interface by deploying this branch.
3. Attempt questions 1 and 2, making use of the autotester to test the results.
4. Attempt questions 3, 4 and 5, saving the chosen answer. Results should be saved to the repository.

To test the edit functionalities:
1. Clone the following repository: https://github.com/githubclassroomtest24032021/sa-tester-test
2. Open the mission in the SA interface by deploying this branch.
3. Additional teacher-only functionality should be available in this branch:
    - the ability to delete the current question
    - the ability to add a new question
    - the ability to delete specific test cases
    - the ability to add new tests cases
    - the ability to delete the 
    - the ability to view and edit the test prepend and test postpend per question
    - the ability to toggle between showing raw text and MCQ display for MCQ questions
3. Use the "add" and "delete" buttons in the control bar to add and delete questions. Save to the repository and check that new questions are created.
4. Click on the task description and the briefing in the side content. Doing so should display the underlying markdown. Make any changes and click away; this should cause the markdown to re-render. Saving to the repository should cause "Problem.md" and "README.md" to change.
5. Attempt to make changes to the code. Saving to the repository should cause edits to "StarterCode.js" and not "SavedCode.js".
6. Navigate to the autotester in the side-content tabs. Add and delete test cases as necessary. Also edit the test programs and expected answers. This should edit or create the file "Testcases.json" in the repository when saved.
7. In the side-content tab, edit the test-prepend and test-postpend. Save to the repository, and see that "TestPrepend.js" and/or "TestPostpend.js" are edited or created.

To test the mission creation functionality:
1. In the Mission Listing page, click on "Create New Mission"
2. Make any changes by
    - Adding questions
    - Editing the briefing
    - Editing the code or description for any question
    - Editing the testcases, test prepend and test postpend for any questions
3. Save the created mission to a new repository.
4. Go back to the mission listing page and load the repository. Check that any saved changes can be loaded.

Issue not fixed:
- Octokit REST API does not allow for repositories to be created for an organisation, only for a user. The ability should exist but this will take a bit of time as we find the new module to add to the project.

### Checklist
- [X] I have tested this code
- [ ] I have updated the documentation

### CODE REVIEW

**Comment by ChengGeng97** [comment]
I'm not sure why it's 541 commits. It should only be about 100. Most of those commits should already have been merged in with P2Minus.

**Comment by ChengGeng97** [comment]
To my knowledge, Octokit API for creating repositories under an organization does not exist for Javascript, only for Ruby.

**Comment by martin-henz** [comment]
> I'm not sure why it's 541 commits. It should only be about 100. Most of those commits should already have been merged in with P2Minus.

I think github counts all commits in this branch from the time it was created. This includes all commits that have been merged from the main branch. That explains the number 541, I think. Don't worry about that.

**Comment by martin-henz** [comment]
> To my knowledge, Octokit API for creating repositories under an organization does not exist for Javascript, only for Ruby.

What is the consequence of this for this project?

File: src/commons/sideContent/githubAssessments/SideContentMissionEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  setMissionMetadata: (missionMetadata: MissionMetadata) => void;
+};
+
+const SideContentMissionEditor: React.FC<SideContentMissionEditorProps> = props => {
+  const datePicker =
+    props.missionMetadata.dueDate.getFullYear() > new Date().getFullYear() ? (
+      <DatePicker onChange={handleDateChange} />
+    ) : (
+      <DatePicker onChange={handleDateChange} defaultValue={props.missionMetadata.dueDate} />
+    );
```
Comment (chownces) [inline src/commons/sideContent/githubAssessments/SideContentMissionEditor.tsx:22]: the calendar UI seems a little off 
image

**Review by chownces — CHANGES_REQUESTED** [review]
Some final touchups!

Btw, is it possible to set a correct answer for the MCQ, and have it light up green when selected?

**Comment by ChengGeng97** [comment]
> > To my knowledge, Octokit API for creating repositories under an organization does not exist for Javascript, only for Ruby.
> 
> What is the consequence of this for this project?

I will make sure to write about this when updating the documentation.

I will also make an issue to track this issue after the merge. I think this is a problem that can be solved by sending our own GET requests to GitHub, perhaps through a similar microservice that is being used for logging-in right now, but making the change now might be cutting a bit too close to the end of our internship.

**Review by chownces — APPROVED** [review]
Looks good! Rmb to file an issue regarding the creation of repositories under an organization. Thanks.

**Pull request merged**